From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NotStevy <notstevy@ultrabuildmc.de>
Date: Fri, 8 Dec 2023 20:58:01 +0100
Subject: [PATCH] Fix versioning with missing group & Update paper


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 29e29fb2228f08dff3edd8bf651220b83b16ed7b..d6110eeabbd5cd6fe101bf602ce9c6ff359d5c84 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/com.infernalsuite.aswm/slimeworldmanager-api/pom.properties"); // ASWM
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/world.kitpvp.aswm/slimeworldmanager-api/pom.properties"); // ASWM
         Properties properties = new Properties();
 
         if (stream != null) {
